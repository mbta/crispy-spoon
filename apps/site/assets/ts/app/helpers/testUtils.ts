/* eslint-disable */
import toJson, { Json } from "enzyme-to-json";
import { ReactWrapper } from "enzyme";
import {
  IndexedRouteStop,
  StopTree
} from "../../schedule/components/__schedule";

export const testRouteStopListFromStopTree = (
  tree: StopTree
): IndexedRouteStop[] =>
  Object.values(tree.byId).map((node, index) => ({
    ...node.value,
    routeIndex: index
  }));

export const createReactRoot = (): void => {
  document.body.innerHTML =
    '<div><div id="react-root"><div id="test"></div></div></div>';
};

export const enzymeToJsonWithoutProps = (wrapper: ReactWrapper<any>): Json =>
  toJson(wrapper, {
    noKey: false,
    map: info => {
      if (!info.props) {
        return info;
      }

      // @ts-ignore is a <style></style> declaration
      if (info.node.type === "style") {
        return info;
      }
      // @ts-ignore is an element in a <noscript></noscript> declaration
      if (info.node.nodeType == undefined) {
        // @ts-ignore
        return info;
      }
      //@ts-ignore is a React component (not HTML element)
      if (info.node.nodeType.toString().includes("function")) {
        // @ts-ignore
        info.props = undefined;
      }

      return info;
    }
  });

export const toJsonEnzyme = (wrapper: ReactWrapper<any>): Json =>
  toJson(wrapper, {
    noKey: false
  });
