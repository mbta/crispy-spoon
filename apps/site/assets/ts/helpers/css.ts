import {
  isABusRoute,
  isACommuterRailRoute,
  isAGreenLineRoute,
  isASilverLineRoute
} from "../models/route";
import { Route } from "../__v3api";

export const routeToCSSClass = (route: string): string =>
  route.toLowerCase().replace(/[ _]/g, "-");

export const busClass = (route: Route): string =>
  isABusRoute(route) && !isASilverLineRoute(route) ? "bus-route-sign" : "";

export const routeToModeName = (route: Route): string => {
  if (isACommuterRailRoute(route)) return "commuter-rail";
  if (route.type === 4) return "ferry";
  if (route.id === "Red" || route.id === "Mattapan") return "red-line";
  if (route.id === "Orange") return "orange-line";
  if (route.id === "Blue") return "blue-line";
  if (isAGreenLineRoute(route)) return "green-line";
  if (isASilverLineRoute(route)) return "silver-line";
  if (isABusRoute(route)) return "bus";
  return "unknown";
};

export const routeBgClass = (route: Route): string =>
  `u-bg--${routeToModeName(route)}`;
