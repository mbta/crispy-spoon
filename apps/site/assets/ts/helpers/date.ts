// Returns a date in the form of "August 7, 2019"
export const formattedDate = (unformatted: string): string => {
  const [year, month, day] = unformatted
    .split(/-/)
    .map(part => Number.parseInt(part, 10));

  // Remember that months in JS are 0-indexed for some reason, hence "month - 1".
  const parsedDate = new Date(year, month - 1, day);
  return parsedDate.toLocaleDateString("en-US", {
    year: "numeric",
    month: "long",
    day: "numeric",
    timeZone: "UTC"
  });
};

// Returns a date in the form of "Aug 7"
export const shortDate = (date: Date): string =>
  date.toLocaleDateString("en-US", {
    month: "short",
    day: "numeric",
    timeZone: "UTC"
  });

export default formattedDate;
