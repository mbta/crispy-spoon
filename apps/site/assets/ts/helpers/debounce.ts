// Debounce is meant to be used with event listeners. Use it to delay a call
// until after a burst of events, such as keyboard input.
//
//   usage:
//
//     const callback = (ev: Event) => {}
//     document.addEventListener("keyup", debounce(callback, 250))
//

const debounce = (callback: Function, delay: number): EventListener => {
  let timer: number | undefined;

  return (...args) => {
    window.clearTimeout(timer);
    timer = window.setTimeout(() => callback(...args), delay);
  };
};

export const debouncePromise = (
  callback: Function,
  delay: number
): Function => {
  let timerId: number | undefined;

  return function debounced(...args: unknown[]) {
    if (timerId) {
      window.clearTimeout(timerId);
    }

    return new Promise(resolve => {
      timerId = window.setTimeout(() => resolve(callback(...args)), delay);
    });
  };
};

export default debounce;
