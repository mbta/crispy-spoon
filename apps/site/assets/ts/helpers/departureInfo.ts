import { concat, difference, keyBy, keys, map, sortBy } from "lodash";
import { PredictionWithTimestamp } from "../models/perdictions";
import { ScheduleWithTimestamp } from "../models/schedules";
import { isCancelled, isDelayed } from "./prediction-helpers";
import { DepartureInfo } from "../models/departureInfo";

const departureInfoToTime = (departureInfo: DepartureInfo): Date => {
  // If there isn't a prediction there should be a schedule
  return departureInfo.prediction
    ? departureInfo.prediction.time
    : departureInfo.schedule!.time;
};

const mergeIntoDepartureInfo = (
  schedules: ScheduleWithTimestamp[],
  predictions: PredictionWithTimestamp[]
): DepartureInfo[] => {
  const schedulesByTripId = keyBy(schedules, sch => sch.trip.id);
  const predictionsByTripId = keyBy(predictions, prd => prd.trip.id);

  const scheduleTripIds = keys(schedulesByTripId);
  const predictionTripIds = keys(predictionsByTripId);

  // All the schedules with and without predictions
  const scheduleDirectionInfo = map(
    schedulesByTripId,
    (schedule, tripId): DepartureInfo => {
      // prediction could be undefined
      const prediction = predictionsByTripId[tripId];
      return {
        prediction: prediction,
        schedule: schedule,
        isCancelled: isCancelled(prediction),
        isDelayed: isDelayed(prediction, schedule)
      };
    }
  );

  const unmatchedPredictionTripIds = difference(
    predictionTripIds,
    scheduleTripIds
  );
  // All the predictions with no schedules
  const predictionDirectionInfo = map(
    unmatchedPredictionTripIds,
    (tripId: string): DepartureInfo => {
      const prediction = predictionsByTripId[tripId];
      return {
        prediction: prediction,
        isCancelled: isCancelled(prediction)
      } as DepartureInfo;
    }
  );

  const departureInfos = concat(scheduleDirectionInfo, predictionDirectionInfo);
  const sortedDepartureInfo = sortBy(departureInfos, di => {
    // prioritize sorting by predictions
    if (di.prediction) {
      return di.prediction.time;
    }
    return di.schedule?.time;
  });

  return sortedDepartureInfo;
};

export { mergeIntoDepartureInfo, departureInfoToTime };
