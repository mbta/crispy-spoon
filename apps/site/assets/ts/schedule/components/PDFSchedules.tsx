import React, { ReactElement } from "react";
import ExpandableBlock from "../../components/ExpandableBlock";
import { SchedulePDF } from "./__schedule";
import renderSvg from "../../helpers/render-svg";
import pdfIcon from "../../../static/images/icon-pdf-default.svg";

const link = (pdf: SchedulePDF): ReactElement<HTMLElement> => (
  <a
    key={pdf.url}
    href={pdf.url}
    rel="noopener noreferrer"
    target="_blank"
    data-turbolinks="false"
    className="m-schedule-page__sidebar-pdfs"
  >
    {renderSvg("c-svg__icon-pdf", pdfIcon)}
    {pdf.title}
  </a>
);

interface Props {
  pdfs: SchedulePDF[];
}

const PDFSchedules = ({ pdfs }: Props): ReactElement<HTMLElement> | null =>
  pdfs.length > 0 ? (
    <ExpandableBlock
      header={{ text: "PDF Schedules and Maps", iconSvgText: null }}
      initiallyExpanded
      id="pdfs"
    >
      <>{pdfs.map(pdf => link(pdf))}</>
    </ExpandableBlock>
  ) : null;

export default PDFSchedules;
