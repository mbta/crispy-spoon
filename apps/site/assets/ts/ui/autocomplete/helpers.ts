import { omit } from "lodash";
import { OnSubmitParams } from "@algolia/autocomplete-core";
import { OnStateChangeProps } from "@algolia/autocomplete-js";
import {
  ContentItem,
  Item,
  LocationItem,
  PopularItem,
  RouteItem,
  StopItem
} from "./__autocomplete";
import { isLGDown } from "../../helpers/media-breakpoints";

export function isStopItem(x: Item): x is StopItem {
  return Object.keys(x).includes("stop");
}

export function isRouteItem(x: Item): x is RouteItem {
  return Object.keys(x).includes("route");
}

export function isContentItem(x: Item): x is ContentItem {
  return Object.keys(x).includes("_content_type");
}

export const getTitleAttribute = (item: Item): string[] => {
  if (isStopItem(item)) {
    return ["stop", "name"];
  }
  if (isRouteItem(item)) {
    return ["route", "name"];
  }

  // @ts-ignore
  // eslint-disable-next-line no-underscore-dangle
  return Object.keys(item._highlightResult);
};

const navStateChange: (props: OnStateChangeProps<Item>) => void = ({
  state
}) => {
  // grey out the page and disable scrolling when search is open
  if (isLGDown()) {
    if (state.isOpen) {
      document.documentElement.dataset.navOpen = "true";
    } else {
      delete document.documentElement.dataset.navOpen;
    }
  }
};

export const STATE_CHANGE_HANDLERS: Record<
  string,
  (props: OnStateChangeProps<Item>) => void
> = {
  nav: navStateChange
};

export const SUBMIT_HANDLERS: Record<
  string,
  (props: OnSubmitParams<Item>) => void
> = {
  to_search_page({ state }) {
    window.Turbolinks.visit(`/search?query=${state.query}`);
  }
};

export type WithUrls<T> = T & { urls: Record<string, string> };

// The backend returns all possible URLs, use urlType to get the desired one
export const itemWithUrl = (
  initialItem: WithUrls<LocationItem | PopularItem>,
  urlType: string
): Omit<typeof initialItem, "urls"> => {
  const item = omit(initialItem, "urls");
  item.url = initialItem.urls[urlType];
  return item;
};
