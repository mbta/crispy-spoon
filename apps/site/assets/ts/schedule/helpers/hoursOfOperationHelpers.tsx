import React, { ReactElement } from "react";
import renderFa from "../../helpers/render-fa";
import { SchedulePDF } from "../components/__schedule";

const pdfLink = (pdf: SchedulePDF | null): ReactElement<HTMLElement> | null => {
  if (!pdf) {
    return null;
  }
  return (
    <div className="fs-16 pt-16">
      <a
        key={pdf.url}
        href={pdf.url}
        rel="noopener noreferrer"
        className="c-call-to-action"
        target="_blank"
      >
        {`Open full schedule and map PDF`}{" "}
        {renderFa("fa-arrow-up-right-from-square", "")}
      </a>
    </div>
  );
};

export default pdfLink;
