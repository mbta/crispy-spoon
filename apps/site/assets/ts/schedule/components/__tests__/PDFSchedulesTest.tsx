import React from "react";
import renderer from "react-test-renderer";
import { createReactRoot } from "../../../app/helpers/testUtils";
import PDFSchedules from "../PDFSchedules";

const pdfs = [
  {
    url: "https://mbta.com/example-pdf.pdf",
    title: "Route 1 schedule PDF"
  }
];

it("it renders", () => {
  createReactRoot();
  const tree = renderer.create(<PDFSchedules pdfs={pdfs} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("it renders without pdfs", () => {
  createReactRoot();
  const tree = renderer.create(<PDFSchedules pdfs={[]} />).toJSON();
  expect(tree).toMatchSnapshot();
});
