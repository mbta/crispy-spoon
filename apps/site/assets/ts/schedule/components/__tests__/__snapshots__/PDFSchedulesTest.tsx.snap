// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
Array [
  <h3
    aria-controls="panel-pdfs"
    aria-expanded={true}
    className="c-expandable-block__header "
    id="header-pdfs"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="c-expandable-block__header-text"
    >
      PDF Schedules
      <span
        className="fa c-expandable-block__header-caret"
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#xF106;",
          }
        }
      />
    </div>
  </h3>,
  <div
    aria-labelledby="header-pdfs"
    className="c-expandable-block__panel "
    id="panel-pdfs"
    role="region"
    tabIndex={0}
  >
    <a
      className="m-schedule-page__sidebar-pdfs"
      data-turbolinks="false"
      href="https://mbta.com/example-pdf.pdf"
      rel="noopener noreferrer"
      target="_blank"
    >
      <span
        aria-hidden="true"
        className="notranslate c-svg__icon-pdf"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
      Route 1 schedule PDF
    </a>
  </div>,
  <noscript>
    <style>
      #header-pdfs { display: none; }
    </style>
    <h3
      className="c-expandable-block__header"
    >
      PDF Schedules
      <span
        className="fa c-expandable-block__header-caret"
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#xF106;",
          }
        }
      />
    </h3>
    <div
      className="c-expandable-block__panel"
      role="region"
    >
      <a
        className="m-schedule-page__sidebar-pdfs"
        data-turbolinks="false"
        href="https://mbta.com/example-pdf.pdf"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span
          aria-hidden="true"
          className="notranslate c-svg__icon-pdf"
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
        Route 1 schedule PDF
      </a>
    </div>
  </noscript>,
]
`;

exports[`it renders without pdfs 1`] = `null`;
